src/core/interaction/pickingcontroller.cpp

```diff
@@ -91,100 +91,106 @@ void PickingController::propagateEvent(Event* event, EventPropagator* propagator
 void PickingController::propagateEvent(TouchEvent* e, EventPropagator* propagator) {
     auto& touchPoints = e->touchPoints();
 
-    std::unordered_map<size_t, std::vector<TouchPoint>> pickngIdToTouchPoints;
-    std::unordered_map<size_t, const PickingAction*> pickingIdToAction;
     std::vector<int> usedPointIds;
 
-    for (auto& point : touchPoints) {
-        const auto coord = glm::clamp(point.pos(), dvec2(0.0), dvec2(e->canvasSize() - uvec2(1)));
-        if (point.state() == TouchState::Started) {
-            auto pa = findPickingAction(coord);
-            tstate_.pointIdToPickingId[point.id()] = pa;
-            tstate_.pickingIdToPressNDC[pa.index] = point.ndc();
+    const auto pickId = [&](size2_t coord) -> size_t {
+        if (auto src = src_.lock()) {
+            const auto value = src->readPixel(coord, LayerType::Picking);
+            if (value.a > 0.0) {
+                return PickingManager::colorToIndex(uvec3(value));
+            }
         }
+        return 0;
+    };
 
-        auto it = tstate_.pointIdToPickingId.find(point.id());
-        auto pa = it != tstate_.pointIdToPickingId.end() ? it->second : findPickingAction(coord);
+    // Strategy for treating touch points for picking objects
+    //
+    // One touch event per initially picked object.
+    //  + Single picking event per object
+    //  + Picked objects will have access to multiple touch points
+    //  - Points initially picking an object may split to multiple object: tricky to maintain state
+    //  (ignored here)
+    //
+    // Other alternative considered:
+    // Treat each touch point individually.
+    //  + Easy to track state
+    //  - Multiple picking events (Enter/Leave/Press) per object possible
+    //  - Picked objects will not have access to multiple touch points
+    //
+
+    // Strategy: One touch event per initially picked object.
+    std::unordered_map<size_t, std::vector<TouchPoint>> pickingIdToTouchPoints;
 
-        if (pa.action) {
-            pickngIdToTouchPoints[pa.index].push_back(point);
-            pickingIdToAction[pa.index] = pa.action;
+    for (auto& point : touchPoints) {
+        if (auto it = touchPointStartPickingId_.find(point.id());
+            it == touchPointStartPickingId_.end()) {
+            // Should mean that point.state() == TouchState::Started
+            const auto coord =
+                glm::clamp(point.pos(), dvec2(0.0), dvec2(e->canvasSize() - uvec2(1)));
+            touchPointStartPickingId_[point.id()] = pickId(coord);
```

Comment:
Why do you clamp here? In case of leaving the window, not clamping will allow you to continue dragging etc. In particular if you are interested in the delta movement since the last event.

Reply:
I think that it is just to make sure that we do not sample the picking id outside of the image. 
@petersteneteg  Can you verify?

A better way would be to return 0 picking id in that case

Reply:
Just to double check: 
We clamp based on `e->canvasSize()`, which may not be the same as `PickingController::src_`?

Reply:
the relevant size it that of the texture we sample I guess...

Reply:
Ok, just wanted to make sure that it did not have anything to do with layout and such